### ISSUE TITLE
wierd behavior with showHelp - I'm losing all the help info...

### ISSUE THREAD

**Issue by osher** [yargs/yargs#614]
I observed a weird collision with typed hyphened args and commands.

## When a typed hyphened argument does not get the type it expects, AND a command is present - yargs.showHelp() behaves in a broken way.

consider this:

```
$ node bin\calc -a 23 -b s4
Synopsis: calc -a <cmd> [num] -b [num]

Commands:
  add  adds 2 numbers
  sub  substracts arg2 of arg1
  mul  multiplies 2 numbers
  div  divides arg1 in arg2

Options:
  -a, --arg1       the first argument                        [number] [required]
  -b, --arg2       the second argument                       [number] [required]
  -p, --precision  when dividing - how many digits to keep              [number]

Error: arguments must be numeric

```

But:

```
C:\ws\sandbox\calc-cli>node bin\calc -a 23 -b s4 add
bin\calc add

Error: arguments must be numeric
```

WTF? where is all the info?

I was passing a lesson about CLI tools in nodeJS, I demonstrated `yargs` as my favorite tool for the job, and I got the wierd behavior in front of all the class.  :stuck_out_tongue: 

The `bin/calc` shell context file:
```
var debug = require('debug')('calc:cli');
var args  = require('../lib/args')(process, console.log);
var calc  = require('../');
var result;

debug('cmd : ', args.cmd);
debug('arg1: ', args.a);
debug('arg2: ', args.b);

try { 
    result = calc.exec( args );
}catch(e) {
    result = e.message
}

console.log("result:\n  %s", result)
```

The `lib/args.js` file:
```
var debug = require('debug')('calc:args');
var assert = require("assert");
var yargs = require('yargs');

module.exports = function(process, log) { 
      
    yargs.reset()
      .usage('Synopsis: calc -a <cmd> [num] -b [num]')
      .command('add', 'adds 2 numbers')
      .command('sub', 'substracts arg2 of arg1')
      .command('mul', 'multiplies 2 numbers')
      .command('div', 'divides arg1 in arg2')
      .options({
        a: {
          alias:      "arg1", 
          describe:   "the first argument",
          type:       'number'
        },
        b: {
          alias:      "arg2", 
          describe:   "the second argument",
          type:       'number'
        }, 
        p : {
          alias:      "precision",
          describe:   "how many digits right to the decimal point to keep",
          type:       'number'
        }
      })
      .demand(['a','b']);
      
    var args = yargs.parse(process.argv);
   
    debug('verifying' , args);
    if (  !isNum(args.a) 
       || !isNum(args.b)
       )  {
        yargs.epilog('Error: arguments must be numeric');
        yargs.showHelp(log);
        process.exit();
        return args
    }

    if (args._.length != 3 ) {
        yargs.epilog('Error: cmd is missing');
        yargs.showHelp(log);
        process.exit();
        return args
    }
    args.cmd = args._.pop();
    
    debug("returning args: ", args)
    
    return args;
    
    function isNum(v) {
        return !isNaN(v) && 'number' === typeof v
        return !isNaN(v) && 'number' === typeof v
    }
}
```

and the `lib/index` calc module (which is the main in package.json):

```
var calc = module.exports = { 
    add: function add(a,b,p) { 
        return $p( a + b )
    },
    sub: function sub(a,b,p) { 
        return $p( a - b )
    },
    mul: function mul(a,b,p) { 
        return $p( a * b )
    },
    div: function div(a,b,p) {
        return $p( a/b, p)
    },
    exec: function exec(args) { 
        if (!calc[args.cmd]) 
            throw new Error('no such command: ' + args.cmd);
        
        return calc[args.cmd](args.a, args.b, args.p);
    }
}

function $p(v, p) {
    if ('number' != typeof p) p = 3;
    p = Math.pow(10,p)
    return Math.round( v * p ) / p
}
```

there are also test files .... but I believe that's enough...

**Comment by nexdrew** [comment]
@osher This is because there was a breaking change made in yargs 5 concerning commands without a builder object or function. Note this item in the [yargs 5.0.0 blog post](http://yargs.js.org/2016/08/14/yargs-5.html):

> builder now defaults to noop builder: the builder now resets all yargs configuration settings that are not flagged as being global.

This means commands *without an explicit builder* are now treated the same as commands with one, resetting option configuration for "command execution". This also means you **must** define top-level options as `global` if you want them to apply when running commands.

The rationale is that yargs now interprets and "executes" all commands equally, regardless of how they're defined. We removed the concept of "documentation-only" commands, such that there's no longer a difference between `.command('foo', 'Description')` and `.command('foo', 'Description', {})`.

So, in your particular example, you should mark your options in `lib/args.js` as `global`, either by adding this one line:

```js
.global(['a', 'b', 'p'])
```

or by using `global: true` for each, like so:

```js
.options({
  a: {
    alias:      "arg1",
    describe:   "the first argument",
    type:       'number',
    global:     true
  },
  b: {
    alias:      "arg2",
    describe:   "the second argument",
    type:       'number',
    global:     true
  },
  p : {
    alias:      "precision",
    describe:   "how many digits right to the decimal point to keep",
    type:       'number',
    global:     true
  }
})
```

Note that, when executing a command, the help text will be command-specific, e.g.

```console
node ./bin/calc.js -a 23 -b s4 add
bin/calc.js add

Options:
  -a, --arg1       the first argument                        [number] [required]
  -b, --arg2       the second argument                       [number] [required]
  -p, --precision  how many digits right to the decimal point to keep   [number]

Error: arguments must be numeric
```

Sorry for the confusion/embarrassment! We took the opportunity of a major version bump to try and make things more consistent, allowing yargs to add more features to commands. Hope this helps!

P.S. Also note that, if you wish, you can now use [coercion](http://yargs.js.org/docs/#methods-coercekey-fn) to do your `isNum()` validation, like so:

```js
.coerce(['a', 'b'], function (v) {
  if (isNaN(v) || typeof v !== 'number') throw new Error('Error: arguments must be numeric')
})
```

**Comment by bcoe** [comment]
@osher feel free to reopen this if @nexdrew's tips don't do the trick 👍 would also love any thoughts you have about how to clarify this further in the README, or on the website.

**Comment by osher** [comment]
Sorry for the late response, I was swapped to do some real coding work, and had to leave the classroom behind, and now that I'm back I can see where all this leads.

So if I get it right - the options model now is broken to global model, and model per command. Very interesting. Now this gets philosophic a little...
Basically - it's a leap of complexity that was not there before. I think the best way to convey it is to simply explain that commands have their own cascading model. 
That makes automating the help output a bit of a challenge more than the challenge to explain the API.

What if an option is demanded for one command, and not the other? What if options need different types for different commands?
I have to play with it a little and see what gets effectively to be your take on this.

Comments I could already provide:

I hated the need to provide the `global: true` flag.
I would suggest that options described before commands should be global by default, and options that are described after a command should be local by default. 

e.g:
```
yargs
. usage("synopsis....", { /* global options */ )
. demand([ /* demanded globally */ ])
.command("cmd1", "descr cmd 1")
.demand([ /* additional mandatory  for cmd 1*/ ])
.options({ /* cascading options for cmd1 */ })
.command("cmd1", "descr cmd 2")
.demand([ /* additional mandatory  for cmd 1*/ ])
.options({ /* cascading options for cmd1 */ })
```

Now this gets clunky. I always prefer the declarative mode providing a data-structure then calling more and more methods. 
So in that sense the next leap would be:

```
yargs
  .usage("synopsis...", { /* global options */ }  )
  .commands( {
     cmd1: { 
       describe: "descr cmd 1",
       options: { /* local options */ },
       handler: fHandleCmd /* func(args) */
     }
  }
```

I would also suggest to support the key `commands` on the global options instead of requiring a different method call, proposing that if you need a switch called `commands` - then give it in the alias, but here I'm not sure how straight forward that is. 

Basically, I aspire to a situation where looking on the data-structure is self explanatory. I think it's simpler and clearer than a set of methods you have to read while compiling the resulting model in the back of your head.

I hope that's helful in any way... :wink:
